futures-util/src/try_future/unwrap_or_else.rs

```diff
@@ -0,0 +1,44 @@
+use core::marker::Unpin;
+use core::mem::PinMut;
+use futures_core::{Future, TryFuture, Poll};
+use futures_core::task;
+
+/// Future for the `unwrap_or_else` combinator. It unwraps the result, returning
+/// the content of the `Ok` as `Output` or if the value is an `Err` then it
+/// calls `op` with its value.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless polled"]
+pub struct UnwrapOrElse<Fut, F> {
```

Comment:
This was previously called `Recover`. The made sense for futures 0.1 and 0.2, but now it works *exactly* like `Result::unwrap_or_else`

Reply:
As another data point, Scala futures call this operation `recover`. That may be where the name came from here. I think I prefer `recover` to `unwrap_or_else`. Especially since it's not actually unwrapping...

Reply:
Its behavior changed in futures 0.3. It now transforms a `Future<Output = Result<T, E>>` into a `Future<Output = T>`. According to git blame this seems to have been defined by @aturon approximately 3 months ago. In think we should align with libstd terminology because we've got an exact match here. Rust users will appreciate it. It is also fitting to change the name because it works now differently than in 0.2.

Reply:
I took a look at Scala's `recover`. Scala futures seem to have a success and error state like we had in futures 0.2 or like in JavaScript.

BTW the only thing that bugs me is that for `TryFuture` the `and_then` and `or_else` methods, which `std::result::Result` both also has, are asynchronous, the others `map_ok` (called `map` on `Result`), `map_err` and now `unwrap_or_else` have synchronous callbacks. In futures 0.2 this was partly sidestepped by implementing `IntoFuture` on `Result`. That can't be done in 0.3. It would work for these `TryFuture` functions, but it wouldn't work on `FutureExt::then` because the return value of the callback can be anything.

Reply:
> As another data point, Scala futures call this operation recover. That may be where the name came from here. I think I prefer recover to unwrap_or_else. Especially since it's not actually unwrapping...

@seanmonstar Yeah, that is where I got the name `recover` from. I don't have a strong opinion here-- I personally think `recover` is clearer, but the symmetry of `unwrap_or_else` with `Result` is nice.